core/bentley/src/Assert.ts

```diff
@@ -17,7 +19,55 @@
  * which indicate a misuse of the API which should be eliminated during development.
  * @public
  */
-export function assert(condition: boolean, msg?: string): asserts condition {
-  if (!condition)
-    throw new Error(`Assert: ${msg ?? "Programmer Error"}`);
+
+/** Asserts that a condition is `true` and - in development builds - throws an error if it is not.
+ * Assertions exist solely to assist programmers during development, in the following ways:
+ *  1 They allow the programmer to declare conditions that they believe cannot possibly occur. If such conditions occur, they indicate
+ *    a serious flaw in the programmer's logic.
+ *  2 They allow the programmer to assure the TypeScript compiler of the truth of some condition that the compiler cannot itself infer.
+ *  3 They allow the author of an API to indicate to consumers of the API a serious misuse that should be corrected during development.
+ *
+ * Assertions should **never** be used to test for conditions - however unlikely - that could be expected to occur at run-time,
+ * such as failing to write to a file or load a resource over the network. If the condition asserted ever fails in a production environment,
+ * the programmer has made a serious mistake.
+ *
+ * By default, `assert` does nothing unless the environment variable `NODE_ENV` exists and is set to `development`. Use [[setAssertionsEnabled]] to override this behavior.
+ *
+ * If your condition and/or message is relatively expensive to compute (e.g., more complex than just a plain boolean or string), pass them as functions to prevent them
+ * from being evaluated when assertions are disabled.
+ *
+ * @param condition The condition that is asserted to be `true`.
```

Comment:
or a function that evaluates to the condition.  The advice paragraph above about when to use a function should be part of this @param doc.

Reply:
Done.